kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerManager.java

```diff
@@ -274,33 +303,107 @@ public String createConsumer(String group, ConsumerInstanceConfig instanceConfig
+                if (!(e instanceof RestException)) {
+                  responseException = Errors.kafkaErrorException(e);
+                }
+                callback.onCompletion(null, responseException);
+              } else {
+                callback.onCompletion(records, null);
               }
-              callback.onCompletion(null, responseException);
-            } else {
-              callback.onCompletion(records, null);
             }
           }
-        }
     );
+    return executor.submit(new RunnableReadTask(new ReadTaskState(task, state, callback)));
+  }
+
+  class RunnableReadTask implements Runnable, Delayed {
+    private final ReadTaskState taskState;
+    private final KafkaRestConfig consumerConfig;
+    private final long started;
+    private final long requestExpiration;
+    // Expiration if this task is waiting, considering both the expiration of the whole task and
+    // a single backoff, if one is in progress
+    private long waitExpiration;
+
+    public RunnableReadTask(ReadTaskState taskState) {
+      this.taskState = taskState;
+      this.started = config.getTime().milliseconds();
+      this.consumerConfig = taskState.consumerState.getConfig();
+      this.requestExpiration = this.started
+              + consumerConfig.getInt(KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      this.waitExpiration = 0;
+    }
+
+    @Override
+    public void run() {
+      try {
+        log.trace("Executing consumer read task ({})", taskState.task);
+        long backoffTime = config.getTime().milliseconds()
+                + consumerConfig.getInt(KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
+
+        taskState.task.doPartialRead();
+        if (!taskState.task.isDone()) {
+          waitExpiration = Math.min(backoffTime, requestExpiration);
+          if (delayedTasks.get(this) != null && delayedTasks.get(this)) {
+            // task is already enqueued
+            return;
+          }
+
+          // add to delayedReadTasks so the scheduler thread can re-schedule another partial read
+          delayedTasks.put(this, true);
+          delayedReadTasks.add(this);
+        } else {
+          log.trace("Finished executing consumer read task ({})", taskState.task);
+        }
+      } catch (Exception e) {
+        log.error("Failed to read records consumer "
+                        + taskState.consumerState.getId().toString(),
+                e);
+        Exception responseException = e;
+        if (!(e instanceof RestException)) {
+          responseException = Errors.kafkaErrorException(e);
+        }
+        taskState.callback.onCompletion(null, responseException);
+      }
+    }
+
+    @Override
+    public long getDelay(TimeUnit unit) {
+      return waitExpiration - config.getTime().milliseconds();
+    }
+
+    @Override
+    public int compareTo(Delayed o) {
+      if (o == null) {
+        return 1;
+      }
+      if (taskState.task.isDone()) {
```

Comment:
This seems like overkill. If we're concerned about accidental submission, maybe we could check for the task completion prior to running? If we happen to hit this, it would likely crash the scheduler thread. I would probably just remove it and rely on our logic to properly check for completion.

Reply:
Yeah it was just sprinkled on top for sanity's sake